GREYErrorFormatter.m

```diff
@@ -0,0 +1,200 @@
+    [logger addObject:hierarchy];
+    return [logger componentsJoinedByString:@"\n"];
+}
+
+#pragma mark - Init
+
+- (instancetype)initWithError:(GREYError *)error {
+  self = [super init];
+  if (self) {
+    _error = error;
+  }
+  return self;
+}
+
+#pragma mark - Public Methods
+
+- (NSString *)formattedDescription {
+  if ([_error.domain isEqualToString:kGREYInteractionErrorDomain] &&
+      _error.code == kGREYInteractionElementNotFoundErrorCode) {
+    return [self elementNotFoundDescription];
+  }
+  
+  // The error code is not yet supported by GREYErrorFormatter
+  return [GREYObjectFormatter formatDictionary:[_error grey_descriptionDictionary]
+                                        indent:kGREYObjectFormatIndent
+                                     hideEmpty:YES
+                                      keyOrder:nil];
+}
+
++ (NSArray *)grey_nestedErrorDictionariesForError:(NSError *)error {
+  if (!error) {
+    return nil;
+  }
+
+  NSMutableArray *errorStack = [[NSMutableArray alloc] init];
+  NSError *underlyingError = error.userInfo[NSUnderlyingErrorKey];
+  if (underlyingError) {
+    NSArray *errorDescriptions = [GREYErrorFormatter grey_nestedErrorDictionariesForError:underlyingError];
+    [errorStack addObjectsFromArray:errorDescriptions];
+  }
+
+  NSDictionary *descriptions = [error grey_descriptionDictionary];
+  // For GREYError, we need to remove some of the fields.
+  if ([error isKindOfClass:[GREYError class]]) {
+    NSMutableDictionary *mutableDescriptions = [descriptions mutableCopy];
+    [mutableDescriptions removeObjectForKey:kErrorUserInfoKey];
+    [mutableDescriptions removeObjectForKey:kErrorErrorInfoKey];
+    [mutableDescriptions removeObjectForKey:kErrorStackTraceKey];
+    [mutableDescriptions removeObjectForKey:kErrorAppUIHierarchyKey];
+    [mutableDescriptions removeObjectForKey:kErrorAppScreenShotsKey];
+    descriptions = mutableDescriptions;
+  }
+  [errorStack addObject:descriptions];
+
+  return errorStack;
+}
+
++ (NSString *)grey_nestedDescriptionForError:(NSError *)error {
+  NSArray *descriptions = [GREYErrorFormatter grey_nestedErrorDictionariesForError:error];
+  if (descriptions.count == 0) {
+    return @"";
+  }
+
+  NSArray *keyOrder = @[
+    kErrorDescriptionKey, kErrorDomainKey, kErrorCodeKey, kErrorFileNameKey, kErrorFunctionNameKey,
+    kErrorLineKey, kErrorTestCaseClassNameKey, kErrorTestCaseMethodNameKey
+  ];
+
+  return [GREYObjectFormatter formatArray:descriptions
+                                   indent:kGREYObjectFormatIndent
+                                 keyOrder:keyOrder];
+}
+
+#pragma mark - Private Methods
+
+- (NSString *)elementNotFoundDescription {
+  NSMutableArray<NSString *> *logger = [[NSMutableArray alloc] init];
+  
+  // exception reason
+  [logger addObject:[NSString stringWithFormat:@"\n%@\n", _error.localizedDescription]];
+
+  // recovery suggestion
+  if (_error.userInfo[kErrorDetailRecoverySuggestionKey]) {
```

Comment:
Store the value to a local variable and reuse that instead of fetching it twice.